### PR TITLE
Fix crash in print_notice & resume reading randfile on signal interrupt

### PR DESCRIPTION
**Pull request by ciz**
Our customers regularly observe this problem with ssh connections:
sshd: fatal: cannot read from /dev/urandom, Interrupted system call
The cause of this is that RAND_load_file() doesn't resume reading from randfile, when it's interrupted by a signal before it managed to read anything.

print_notice() doesn't check i2s_ASN1_INTEGER's return value for NULL, so the invalid pointer is later passed to strlen() where it causes a crash.
This was encountered by browsing www.digikam.org with w3m.

### CODE REVIEW

**Comment by richsalz** [comment]
+1, but we need a CLA or a comment from you that you consider this to be trivial.

**Comment by mattcaswell** [comment]
Drive by: This doesn't look trivial to me. I think it needs a CLA.

File: crypto/rand/randfile.c [hunk 1]
```diff
@@ -162,7 +162,13 @@ int RAND_load_file(const char *file, long bytes)
             n = (bytes < BUFSIZE) ? (int)bytes : BUFSIZE;
         else
             n = BUFSIZE;
+
         i = fread(buf, 1, n, in);
+        if (i <= 0 && ferror(in) && errno == EINTR) {
```
Comment (FdaSilvaYY) [inline crypto/rand/randfile.c]: As EINTR is guarded by #if in the current codebase, shouldn't this change follow the pattern ?

Comment (richsalz) [inline crypto/rand/randfile.c]: yes, good point.  Maybe something like
```
#ifdef EINTR
#define INTERRUPTED(in) ((ferror(in) && errno == EINTR)
#else
#define INTERRUPTED (0) /* or ferror? */
#endif
```

Comment (ciz) [inline crypto/rand/randfile.c]: I didn't realize that other usages of EINTR are hidden behind ifdefs. Thanks for pointing that out.
I've updated the pull request, based on Rich's suggestion.

**Comment by ciz** [comment]
@mattcaswell, I asked our legal guys about signing CCLA.

**Comment by ciz** [comment]
I've just sent a signed ICLA.

**Comment by ciz** [comment]
The pull request has been updated with the #ifdefs around EINTR as suggested by FdaSilvaYY and richsalz. Does the patch set look fine now?

File: crypto/rand/randfile.c [hunk 2]
```diff
@@ -162,9 +168,17 @@ int RAND_load_file(const char *file, long bytes)
...
+
         i = fread(buf, 1, n, in);
-        if (i <= 0)
-            break;
+        if (i <= 0) {
+            if (INTERRUPTED(in)) {
+                /* Interrupted by a signal, resume reading */
+                clearerr(in);
+                continue;
+            } else {
```
Comment (richsalz) [inline crypto/rand/randfile.c]: one last thing.  get rid of the else after the continue.

**Comment by ciz** [comment]
I broke the pull request, when attempting to squash the commits, sorry :-(
I'll open a new one, unless I figure out how to fix this.

**Comment by richsalz** [comment]
It wouldn't be git if that didn't happen once in a while.  no worries :)

**Comment by ciz** [comment]
This pull request is superseded by https://github.com/openssl/openssl/pull/2044 and https://github.com/openssl/openssl/pull/2045

**Pull request closed without merging**